detectors/resources/src/main/java/com/google/cloud/opentelemetry/detectors/GCFResource.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2022 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.opentelemetry.detectors;
+
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;
+
+public final class GCFResource implements ResourceProvider {
```

Comment:
I see what you mean here, yes they are all kind of dependent on each other in the sense that ultimately all enabled resource providers are merged into a single resource internally - which is then disambiguated by ResourceTranslator.  

I suppose it should be possible to combine all of them in a single GCPResourceDetector class, I am not sure how collector is implemented, but IMO it really is a question of readability. The way it's currently structured makes it clear that what attributes are required to identify an individual platform. 

If I am understanding this correctly, then if we combine all of them in a single class -  the class will try to identify and put all known attributes in a single attributesBuilder and then differentiate them towards the end through the unique attribute from the spec document. 

This might also make testing a little difficult (currently we are able to assert number of attributes in individual classes, but I imagine if we combine, all of them will have same number of attributes).

Reply:
Here is the collector's implementation, for reference: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/gcp/gcp.go#L59

The implementation of those functions is here: https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/tree/main/detectors/gcp

Reply:
Isn't this similar to the current implementation in Java ? 

The `Detect` method in collector contrib is like the [mapResource](https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/blob/main/shared/resourcemapping/src/main/java/com/google/cloud/opentelemetry/resource/ResourceTranslator.java#L106) method and the individual implementations linked above are what Java's current  [detectors](https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/tree/main/detectors/resources/src/main/java/com/google/cloud/opentelemetry/detectors) look like.

Reply:
Is there a single public function/class for customers that discovers for the all platforms, or would I need to call each platform individually?

Reply:
I dont think you need to call individual platform individually as a customer. 

From the customer's point of view in order to detect what platform it is they just need to call `getAttributes().get(ResourceAttributes.CLOUD_PLATFORM)` on the current `Resource`.  

For Java, we can get the Resource from the `AutoConfiguredOpenTelemetrySdk` and this resource basically contains a combination of all the resource providers defined and are disambiguated through the mappings we provide in the `mapResource` method linked above.

Let me know if this is what you were asking.

Reply:
In `io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider`, you list all the detectors that are installed when using autoconfigured resource detectors.  You list one detector for each platform.

It might be better if there was just one element there: `com.google.cloud.opentelemetry.detectors.GCPResource`, instead of one for each platform.  It might be a good idea to unify these in a separate PR.

Feel free to come find me today, and we can chat if you still have questions.

Reply:
I see the benefit in unifying is a simplification in setting up the `OpenTelemetrySdk` (unless we are using `AutoConfiguredOpenTelemetrySdk`). 

```java
      Resource resource = Resource.getDefault()
              .merge(Resource.create(new GCEResource().getAttributes()))
              .merge(Resource.create(new GCFResource().getAttributes()))
              .merge(Resource.create(new GKEResource().getAttributes()))
              .merge(Resource.create(new CloudRunResource().getAttributes()));
      
      SdkTracerProvider tracerProvider = SdkTracerProvider.builder()
                .addSpanProcessor(BatchSpanProcessor.builder(traceExporter).build())
                .setResource(resource)
                .build();
```

I agree that combining into a single class might help, but I think we might then loose the granularity on what ResourceProviders we can  enable/disable - https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure#resource-provider-spi

Reply:
Created an issue to address this issue - https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/issues/174 

Feel free to add relevant details to this issue.

Reply:
Thanks!